// @headspinio/tizen-remote/build/lib/command.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TextCommand_text;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextCommand = exports.KeyCommand = void 0;
const tizen_remote_1 = require("./tizen-remote");
class KeyCommand {
    /**
     * @param {import('./tizen-remote').KeyCommandType} cmd
     * @param {import('./tizen-remote').RcKeyCode} data
     */
    constructor(cmd, data) {
        this.method = tizen_remote_1.constants.COMMAND_METHOD;
        this.params = {
            Cmd: cmd,
            DataOfCmd: data,
            Option: tizen_remote_1.constants.COMMAND_PARAMS_OPTION,
            TypeOfRemote: 'SendRemoteKey',
        };
    }
}
exports.KeyCommand = KeyCommand;
class TextCommand {
    /**
     * @param {string} text
     */
    constructor(text) {
        this.method = tizen_remote_1.constants.COMMAND_METHOD;
        /**
         * Original text
         * @type {string}
         */
        _TextCommand_text.set(this, void 0);
        __classPrivateFieldSet(this, _TextCommand_text, text, "f");
        this.params = {
            Cmd: Buffer.from(text).toString('base64'),
            DataOfCmd: 'base64',
            Option: tizen_remote_1.constants.COMMAND_PARAMS_OPTION,
            TypeOfRemote: 'SendInputString',
        };
    }
    get text() {
        return __classPrivateFieldGet(this, _TextCommand_text, "f");
    }
}
exports.TextCommand = TextCommand;
_TextCommand_text = new WeakMap();
